File: prob15.js
//Joe McCourt
//4/15/12

// Project Euler problem 15
// Starting in the top left corner of a 22 grid, 
// there are 6 routes (without backtracking) to the 
// bottom right corner.
// How many routes are there through a 2020 grid?
function Prob15(){

	this.eval = function(){
		//Initialize number of paths array (21x21)
		var w = 21;
		var h = 21;
		var numPaths = [];
		for(var y = 0; y<h; y++){
			for(var x = 0; x<w; x++){
				numPaths[y*w+x] = 1;
			}
		}

		//The number of paths for each vertex
		//is the sum of the number of paths on
		//the vertex above and to the vertex left
		for(var y = 1; y<h; y++){
			for(var x = 1; x<w; x++){
				numPaths[y*w+x] = numPaths[y*w+x-1]+numPaths[(y-1)*w+x];
			}
		}

		return numPaths[w*h-1];
	};
}

File: go/prob15.go
// Starting in the top left corner of a 2×2 grid,
// and only being able to move to the right and down,
// there are exactly 6 routes to the bottom right corner.

// How many such routes are there through a 20×20 grid?


// 2x2
// 6 3 1
// 3 2 1
// 1 1 1

// 3x3
// 20 10 4  1
// 10 6  3  1
// 4  3  2  1
// 1  1  1  1


package main

import (
	"fmt"
)

func prob15() {

	n := 20
	grid := make([][]uint64, n+1)
	for i := range grid {
		grid[i] = make([]uint64, n+1)
	}

	for i := n; i >= 0; i-- {
		for j := n; j >= 0; j-- {
			if j == n || i == n {
				grid[i][j] = 1
			} else {
				grid[i][j] = grid[i+1][j] + grid[i][j+1]
			}
		}
	}

	fmt.Println(grid)
	fmt.Println(grid[0][0])
}
